nvaccess/nvda
# Option to configure when the error sound may be played.

### Link to issue number:

#12672

### Summary of the issue:
The error sound that is played in test versions may also be useful to testers on stable and RC releases.
The motivation to have this feature included in NVDA's core rather than in an add-on is that it may be useful to test NVDA stable versions with add-ons disabled.

### Description of how this pull request fixes the issue:

Added a 'Play a sound for logged errors' option in the advanced settings panel. It is a combo-box with two choices:
- 'Only in test versions': NVDA plays error sounds as today, i.e. only in NVDA test versions (alpha, beta and from source) but not in RC or stable.
- Yes: NVDA plays the error sound whatever its version (test or stable)
According to the discussion in #12672, I do not provide an option to disable error sounds in tests versions, since test versions are aimed to capture potential issues. Thus error notifications should not be disabled for this usage.

If an error occurs before the config is initialized, the default behaviour is applied, i.e. the error sound is played only in test versions.

Note for reviewers:
* I have had to import config in the `handle` method rather than at the top of the module since NVDA could not start in the latter case. I do not know why however, since I do not know anymore how to debug this since the virtual environment has been set up.
* Please native English speakers, check my writings (GUI and user guide)

Additional modification:
While working on this part of the code, I have made the critical error sound played asynchronously upon @leonardder's suggestion. It should allow to spare 2-3 seconds (the time of the error sound) in the code execution time.
NV Access, if you prefer not to include this modification for whatever reason, I or you may revert this commit.

### Testing strategy:

Manual testing:
Triggered an error in the 4 following situations and checked that the error sound is played for the 3 first ones and not the last one:
* NVDA test version (from source) and option set to 'Yes'
* NVDA test version (from source) and option set to 'Only in NVDA test versions'
* NVDA stable version (simulated from source) and option set to 'Yes'
* NVDA stable version (simulated from source) and option set to 'Only in NVDA test versions'

The stable version is simulated running NVDA from source with this branch and entering the 2 following lines in the Python console:
```
import buildVersion
buildVersion.isTestVersion = False
```

The error is triggered by trying to enable the screen curtain while Windows Magnifier is enabled with color inversion activated. But many other ways could have been used.

I will also check context help on the snapshot generated by appVeyor.

### Known issues with pull request:
If an error occurs before the config is initialized, the default behaviour is applied, i.e. the error sound is played only in test versions whatever the value of this option is in the config.

### Change log entries:
New features (or Changes? Or For developers?)
It is now possible to configure the error sound to be played in any version of NVDA, not only in test versions. (#12672)

### Code Review Checklist:

  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
